### ISSUE TITLE
[ASSETS] Image shortcode parser 

### ISSUE THREAD

**Issue by cwchong** [silverstripe/silverstripe-assets#178]
## Affected Version
4.1.1

## Description
According to [https://docs.silverstripe.org/en/4/developer_guides/extending/shortcodes/](https://docs.silverstripe.org/en/4/developer_guides/extending/shortcodes/) 

> Shortcodes are automatically parsed on any database field which is declared as HTMLValue or DBHTMLText, when rendered into a template. 

However, declaring a field via `DBHTMLText::class` instead of the previous `HTMLText` does not trigger shortcode parsing as suggested in the documentation.

## Steps to Reproduce
`private static $db = [
        'Content1' => 'HTMLText',
        'Content2' => DBHTMLText::class,
    ];`

Image shortcodes in Content2 will render as unparsed [img] instead of the actual <img> html

**Comment by ivoba** [comment]
Im also having this issue on 4.3 (without CMS) even on HTMLText.
@dhensby Could you explain how to enable this:

> DataObject enables shortcode parsing if type === HTMLText::class

**Comment by robbieaverill** [comment]
@ivoba that is enabled with YAML configuration in the framework module:

```yaml
---
Name: corefieldtypes
---
SilverStripe\Core\Injector\Injector:
  # ...
  HTMLText:
    class: SilverStripe\ORM\FieldType\DBHTMLText
    properties:
      ProcessShortcodes: true
```

This means that when you use DB fields that are `HTMLText` type, they'll automatically have `->setProcessShortcodes(true)` on their instance when instantiated. As @dhensby mentioned, this isn't the same when using `DBHTMLText::class` directly in DB field definitions.

We could either enable this for DBHTMLText via injector configuration as well, or remove that from the documentation.

@martinduparc @ivoba are you both experiencing issues with a DB definition like this (below) not parsing shortcodes correctly?

```php
private static $db = ['MyField' => 'HTMLText'];
```

**Comment by ivoba** [comment]
@robbieaverill yes this coniguration is present however in vendor/silverstripe/framework/src/ORM/FieldType/DBHTMLText.php:132
`$this->processShortcodes` is false

Some Context:
Im not using the Cms, just the ModelAdmin and renderWith in a Controller.
I fixed it now in the Controller by
```
$arrayData = new ArrayData([
                'Title'   => $page->first()->Title,
                'Content' => ShortcodeParser::get_active()->parse($page->first()->Text),
            ]);
```

But still seems not correct that ShortcodeParser has to be applied manually.

**Comment by robbieaverill** [comment]
>this coniguration is present however in vendor/silverstripe/framework/src/ORM/FieldType/DBHTMLText.php:132 $this->processShortcodes is false

@ivoba This would be expected behaviour _currently_ if you're using `DBHTMLText::class` in your DB definition, because Injector can't apply the `(set)ProcessShortcodes: true` call when it instantiates the class. Can you confirm that you're using that, and not `$db = ['Text' => 'HTMLText']` which should work correctly?

**Comment by ivoba** [comment]
@robbieaverill yes im using HTMLText
```
    private static $db = [
        'Title' => 'Varchar',
        'Text'  => 'HTMLText',
    ];
```

**Comment by robbieaverill** [comment]
OK, that looks like a bug then

**Comment by martinduparc** [comment]
I can confirm that `'Text'  => 'HTMLText'` works while `'Text'  => DBHTMLText::class` doesn't

**Comment by aNickzz** [comment]
What was the reason for using these proxy classes in the first place (`'Text'` instead of `DBText::class`)? for nearly all classes its just an alias _except_ `DBHTMLText` which is a one off.

Until we had this problem I had thought that the string literals were just aliases `$actualClass = 'DB' . $dbValue`. Using the proxy classes just seem to be a layer of complication, I would much rather see `DBHHTMLText::class` than `'HTMLText'` as its more direct and has other benefits like goto source in editors with intellisense.

How likely is it that a PR setting `processShortcodes` to true by default (and removing the newly redundant yml property) would be accepted? Would it need to wait till 5.x?

Honestly I would like to see all these alias classes ( https://github.com/silverstripe/silverstripe-framework/blob/01ed8a316b65c7d2dddecec39c0195a0e3de07b2/_config/model.yml#L4-L60 ) disappear altogether unless there's a good reason to keep them.

**Comment by maxime-rainville** [comment]
> How likely is it that a PR setting processShortcodes to true by default (and removing the newly redundant yml property) would be accepted? Would it need to wait till 5.x?

There's no way we could do this change in a minor release. If you upgrade your site and a field that use to not process short code suddenly starts doing so, it would be a breaking change.

> Honestly I would like to see all these alias classes ( _config/model.yml#L4-L60 ) disappear altogether unless there's a good reason to keep them.

I think there's value in keeping the current string aliases for `$db`. The point of the `$db` array is to define what columns you want on your table. How SilverStripe interacts with those columns and manage them is a detail most developers don't really need to know to use the ORM.

Using the `::class` notation for relationships makes sense because those will return actual instances of those classes.

**Comment by aNickzz** [comment]
> There's no way we could do this change in a minor release.

I hear SilverStripe 5 will be around soon, maybe this could be reconsidered.

> How SilverStripe interacts with those columns and manage them is a detail most developers don't really need to know to use the ORM.

I would argue that using the class names explicitly would make the ORM easier to understand. The allowed values in the `$db` array are, at the end of the day, the DBField classes. Using them explicitly would make it easier for newcomers to discover the helper functions available in the templates, as well as what other field types are available for use.

It would also make it simpler to use custom field types (like those that extend `DBComposite`). Using the aliases defined by the injector adds a layer of convolution and sacrifices the IntelliSense you'd get by the `::class` syntax, not to mention that using `::class` already works currently.
